packages/versioning/src/versioning.ts

```diff
@@ -0,0 +1,154 @@
+import { NumericLiteralType, Program, StringLiteralType, Type } from "@cadl-lang/compiler";
+const addedOnKey = Symbol();
+const removedOnKey = Symbol();
+const versionsKey = Symbol();
+const renamedFromKey = Symbol();
+const madeOptionalKey = Symbol();
+
+export function $added(p: Program, t: Type, v: StringLiteralType | NumericLiteralType) {
+  p.stateMap(addedOnKey).set(t, v);
+}
+export function $removed(p: Program, t: Type, v: StringLiteralType | NumericLiteralType) {
+  p.stateMap(removedOnKey).set(t, v);
+}
+export function $renamedFrom(
+  p: Program,
+  t: Type,
+  v: StringLiteralType | NumericLiteralType,
+  oldName: StringLiteralType
+) {
+  const record = { v: v, oldName: oldName };
+  p.stateMap(renamedFromKey).set(t, record);
+}
+
+export function $madeOptional(p: Program, t: Type, v: string | boolean) {
+  p.stateMap(madeOptionalKey).set(t, v);
+}
+
+export function getRenamedFromVersion(p: Program, t: Type) {
+  return p.stateMap(renamedFromKey).get(t)?.v ?? -1;
+}
+export function getRenamedFromOldName(p: Program, t: Type) {
+  return p.stateMap(renamedFromKey).get(t)?.oldName ?? "";
+}
+export function getAddedOn(p: Program, t: Type) {
+  return p.stateMap(addedOnKey).get(t) ?? -1;
+}
+export function getRemovedOn(p: Program, t: Type) {
+  return p.stateMap(removedOnKey).get(t) ?? Infinity;
+}
+
+export function getMadeOptionalOn(p: Program, t: Type) {
+  return p.stateMap(madeOptionalKey).get(t) ?? -1;
+}
+
+export function $versioned(p: Program, t: Type, v: Type) {
```

Comment:
We will want to ensure (probably in a linter added later) that every version in this list contains at least one change (i.e. at least one versioning decorator that uses the version)

Reply:
Is it not valid to add a new version simply because behavior changed?

Reply:
Behavior changed in a way not covered by the API specification?  I suppose that we can add this check as a linting rule for azure.